lib/compat/wordpress-6.2/get-global-styles-and-settings.php

```diff
@@ -224,6 +224,55 @@ function gutenberg_get_global_settings( $path = array(), $context = array() ) {
 		$origin = 'theme';
 	}
 
-	$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin )->get_settings();
+	$cache_group = 'theme_json';
+	$cache_key   = 'gutenberg_get_global_settings_' . $origin;
+	$settings    = wp_cache_get( $cache_key, $cache_group );
+
+	if (
+		false === $settings ||
+		( defined( 'WP_DEBUG' ) && WP_DEBUG )
+	) {
+		$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data( $origin )->get_settings();
```

Comment:
Consider these lines 
https://github.com/WordPress/wordpress-develop/blob/bb948c5f6d462dc43483f198d36e9b0948d2e5f1/src/wp-includes/class-wp-theme-json-resolver.php#L297
get_default_block_editor_settings has calls to get_options for image sizes, `is_rtl` related to langaguge setting of the user, `get_default_block_categories` that is filterable, `get_allowed_mime_types` filterable, and `get_theme_support` that can removed. How are all these cases invalidated?

Reply:
This is great feedback:

- `get_theme_support`, `get_default_block_categories`, `get_allowed_mime_types` are filtered by a consumer: either themes or plugins. We clean the cache when a plugin is activated/deactivated, theme switch, etc.. so these cases are covered.
- `is_rtl`: I need to look at this one. We may need to invalidate the cache upon language switches as well.

Reply:
As noted, updating a plugin could change the code. Also what if someone updates their plugin by uploading it using FTP. Something I have done in the past and is very common. 

Is_rtl can be related to the current user, so can not be cached ever. We need a rethink here.

Reply:
When someone updates something via FTP, this code will work perfectly in the next request if they don't use a persistent cache plugin. If they do, they can always clear the cache.

Reply:
Yes, but currently, there are no other parts of core, that require you to flush caches on FTP uploads. Changing this behavior, would be a breaking change, which is something we need avoid at all costs.

Reply:
> currently, there are no other parts of core, that require you to flush caches on FTP uploads

It's not that needed for themes for instance? Core uses some cached theme data in https://core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L745 and looks like it has the same problem with FTP uploads (they are required to use `WP_Theme::cache_delete` or `wp_clean_themes_cache`).

Reply:
> https://core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L745

This is different theme by default are not a persistent cache group. [See](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L232)

Reply:
> > [core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L745](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L745)
> 
> This is different theme by default are not a persistent cache group. [See](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/class-wp-theme.php?rev=54236#L232)

Here we are dealing with site styles that are related to what a theme does. In that case, shouldn't we just follow the same approach and make sure this cache is not in a persistent cache group?

Reply:
> is_rtl related to langaguge setting of the user, get_default_block_categories that is filterable, get_allowed_mime_types filterable 


Although get_default_block_editor_settings gets this information, it is not used at all for the purpose of computing theme. json related information, so even if filters change this, it does not affect the output of the cached functions.

Reply:
> Although get_default_block_editor_settings gets this information, it is not used at all for the purpose of computing theme. json related information, so even if filters change this, it does not affect the output of the cached functions.

Wow, I never followed that code path. Calling `get_default_block_editor_settings` is wasteful then. We should fix that ASAP. To that end I have created https://github.com/WordPress/gutenberg/pull/46112.

Reply:
In this PR, I worked on adding caching to theme_json_resolver, as had core unit tests fail. https://github.com/WordPress/wordpress-develop/pull/3624. 

This was because of add / removing theme supports. I don't think we should merge this until we have solution to this problem.